process/dns_v1.go

```diff
@@ -0,0 +1,336 @@
+package process
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math"
+
+	"github.com/DataDog/mmh3"
+)
+
+// DNS data is encoded as a very basic bucketed hash table.  There are three blocks, or buffers, of data:
+//
+//	The "name" block is all of the unique DNS names.  The length of the name is stored as a varint
+//	followed by the name itself
+//
+//	The "bucket" block contains all of the hash buckets.  The format of each bucket is:
+//		varint for number of entries in bucket
+//		For each entry in the bucket:
+//			varint for length of key
+//			key bytes
+//			varint for number of names associated with the key
+//			Each associated name is encoded as a varint which is the position of the actual name string in the name block
+//
+//	The "position" block is a list of varints, one for each bucket, where each varint is a pointer to the start
+//	of the bucket in the bucket block
+//
+// The overall buffer is encoded as:
+//	1 byte indicating version
+// 	1 byte indicating the number of buckets
+//	varint indicating the length of the name buffer.
+//	varint indicating the length of the position buffer
+// 	varint indicating the position of the "middle" (bucketCount / 2) bucket in the position block
+//		We will use this to skip the half of the buckets when searching for the target bucket index
+//	position block
+//	bucket block
+//	name block
+//
+// Notes:
+//	Using varints saves space at the cost of not having random access to certain sections of data, particularly the
+//	bucket position mapping.  This was a deliberate trade off to reduce the size of the payload and thus memory usage
+//
+//	Varints are also more finicky to deal with in terms of calculating required space ahead of time.  This increases
+//	the implementation complexity, or at least the line count, but we reduce allocations & memory usage by
+// 	pre-sizing the output buffers
+//
+// This type is not thread safe
+type V1DNSEncoder struct {
+	BucketFactor float64
+	scratch      [binary.MaxVarintLen64]byte // Used for varint encoding
+}
+
+type bucketEntry struct {
+	keys []string
+	size int
+}
+
+// 1 byte for version, 1 byte for bucket count
+const dns1Version1PreambleLength = 2
+
+// Used for calculating the number of buckets for a given input map.
+// Currently the bucket count is calculated as `len(input) * bucketFactor`
+const defaultBucketFactor = 0.75
+
+func NewV1DNSEncoder() DNSEncoder {
+	return &V1DNSEncoder{
+		BucketFactor: defaultBucketFactor,
+	}
+}
+
+func (e *V1DNSEncoder) Encode(dns map[string]*DNSEntry) []byte {
```

Comment:
Curious as to why we opted for dealing with IPs as strings instead of raw bytes.
255.255.255.255 has 15 bytes as opposed to 4 bytes in raw representation (0xFF 0xFF 0xFF 0xFF).
I know system-probe is sending back a `map[string]*DNSEntry` but we can totally change that now that you're putting a serious though on encoding.

Reply:
An earlier prototype of this did in fact use bytes for IP addresses.  They have the nice property of being fixed size so you can skip through them more easily when scanning.

Unfortunately since we deal with the data as strings in the agent and backend, it requires converting them to `net.IP`, which allocates.  This occurs both on encoding an decoding and was expensive enough that it wasn't worth it.

@sunhay had a PR a while back that converted our addresses to a non-string representation but I think we need to think through it a little more because of how we deal with them on the back-end, mainly using them as map keys.